File: backend/utils/excel_export.py
import io
import openpyxl
from openpyxl.utils import get_column_letter
from openpyxl.styles import Alignment, PatternFill, Border, Side, Font
from sqlalchemy.orm import Session
import models

def generate_timetable_excel_bytes(department_code: str, semester: int, db: Session) -> io.BytesIO:
    slots = db.query(models.SlotMaster).order_by(models.SlotMaster.day_of_week, models.SlotMaster.period_number).all()
    if not slots:
        raise ValueError("No slots configured")

    days_order = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
    active_days = [d for d in days_order if any(s.day_of_week == d for s in slots)]
    if not active_days: active_days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]

    max_period = max([s.period_number for s in slots], default=8)
    periods = list(range(1, max_period + 1))

    query = db.query(models.TimetableEntry)
    if department_code: query = query.filter(models.TimetableEntry.department_code == department_code)
    if semester: query = query.filter(models.TimetableEntry.semester == semester)
    entries = query.all()

    timetable_dict = {}
    for e in entries:
        key = (e.day_of_week, e.period_number)
        timetable_dict.setdefault(key, []).append(e)

    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = f"Timetable {department_code or 'All'} Sem {semester or 'All'}"

    thin = Side(border_style="thin", color="D1D5DB")
    border_all = Border(top=thin, left=thin, right=thin, bottom=thin)
    fill_header = PatternFill(start_color="F9FAFB", end_color="F9FAFB", fill_type="solid")
    fill_day = PatternFill(start_color="F9FAFB", end_color="F9FAFB", fill_type="solid")
    fill_lab = PatternFill(start_color="FEF9C3", end_color="FEF9C3", fill_type="solid")
    fill_class = PatternFill(start_color="FFFFFF", end_color="FFFFFF", fill_type="solid")
    fill_empty = PatternFill(start_color="F3F4F6", end_color="F3F4F6", fill_type="solid")
    font_bold = Font(bold=True, color="1F2937", size=11)
    align_center = Alignment(horizontal='center', vertical='center', wrap_text=True)

    header_row = 2
    d_cell = ws.cell(row=header_row, column=1, value="DAY / PERIOD")
    d_cell.font = font_bold
    d_cell.alignment = align_center
    d_cell.fill = fill_day
    d_cell.border = border_all
    ws.column_dimensions['A'].width = 16

    for i, p in enumerate(periods):
        col = i + 2
        slot = next((s for s in slots if s.period_number == p and s.day_of_week == 'Monday'), None)
        time_str = f"{slot.start_time} - {slot.end_time}" if slot else ""
        header_text = f"Period {p}\n({time_str})"
        c = ws.cell(row=header_row, column=col, value=header_text)
        c.font = Font(bold=True, color="374151", size=10)
        c.alignment = align_center
        c.fill = fill_header
        c.border = border_all
        ws.column_dimensions[get_column_letter(col)].width = 24

    current_row = header_row + 1
    for day in active_days:
        dc = ws.cell(row=current_row, column=1, value=day[:3].upper())
        dc.font = Font(bold=True, size=12, color="1F2937")
        dc.alignment = align_center
        dc.fill = fill_day
        dc.border = border_all
        ws.row_dimensions[current_row].height = 85

        for i, p in enumerate(periods):
            col = i + 2
            cell_entries = timetable_dict.get((day, p), [])
            if not cell_entries:
                cell_text = "Empty Slot"
                fill = fill_empty
                cell_font = Font(italic=True, color="9CA3AF", size=10)
            else:
                is_lab = any(e.session_type and e.session_type.upper() == 'LAB' for e in cell_entries)
                lines = []
                for e in cell_entries:
                    block = f"{e.course_code}\n{e.course_name or ''}\n({e.faculty_name or 'Unassigned'})\n[{e.venue_name or ''}]"
                    lines.append(block)
                cell_text = "\n---\n".join(lines)
                fill = fill_lab if is_lab else fill_class
                cell_font = Font(bold=False, size=10, color="111827")

            c = ws.cell(row=current_row, column=col, value=cell_text)
            c.alignment = align_center
            c.fill = fill
            c.border = border_all
            c.font = cell_font
        current_row += 1

    output = io.BytesIO()
    wb.save(output)
    output.seek(0)
    return output

File: backend/utils/excel_export.js
/**
 * excel_export.js
 * 
 * Port of Python's utils/excel_export.py using ExcelJS for full cell styling.
 * Produces a timetable grid with the exact same formatting as the original:
 *   - Header row with period numbers + time ranges
 *   - Day rows with abbreviated day names
 *   - Cells with course_code, course_name, faculty_name, venue_name
 *   - Lab cells highlighted yellow, empty cells grey, normal cells white
 *   - Thin borders, centered alignment, bold headers, wrap text
 */

const ExcelJS = require('exceljs');

/**
 * Generate a formatted timetable Excel buffer.
 * @param {object} db - better-sqlite3 database instance
 * @param {string} department_code
 * @param {number|string} semester
 * @returns {Promise<Buffer>} xlsx buffer ready to send
 */
async function generateTimetableExcelBytes(db, department_code, semester) {
    // Fetch slot master for time ranges
    let slots = [];
    try {
        slots = db.prepare("SELECT * FROM slot_master ORDER BY day_of_week, period_number").all();
    } catch (e) { /* slot_master may not exist */ }

    const daysOrder = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];
    let activeDays = daysOrder.filter(d => slots.some(s => s.day_of_week === d));
    if (!activeDays.length) activeDays = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"];

    const maxPeriod = slots.length > 0 ? Math.max(...slots.map(s => s.period_number)) : 8;
    const periods = Array.from({ length: maxPeriod }, (_, i) => i + 1);

    // Fetch timetable entries
    const entries = db.prepare(
        "SELECT * FROM timetable_entries WHERE department_code = ? AND semester = ? ORDER BY day_of_week, period_number"
    ).all(department_code, parseInt(semester));

    // Build lookup dict: (day, period) -> [entries]
    const timetableDict = {};
    for (const e of entries) {
        const key = `${e.day_of_week}|${e.period_number}`;
        if (!timetableDict[key]) timetableDict[key] = [];
        timetableDict[key].push(e);
    }

    // --- Create workbook ---
    const wb = new ExcelJS.Workbook();
    const ws = wb.addWorksheet(`Timetable ${department_code || 'All'} Sem ${semester || 'All'}`);

    // --- Styles (matching Python openpyxl originals) ---
    const thinBorder = { style: 'thin', color: { argb: 'FFD1D5DB' } };
    const borderAll = {
        top: thinBorder, left: thinBorder, right: thinBorder, bottom: thinBorder
    };
    const fillHeader = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFF9FAFB' } };
    const fillDay = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFF9FAFB' } };
    const fillLab = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFEF9C3' } };
    const fillClass = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFFFFF' } };
    const fillEmpty = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFF3F4F6' } };

    const fontBold = { bold: true, color: { argb: 'FF1F2937' }, size: 11 };
    const fontHeaderPeriod = { bold: true, color: { argb: 'FF374151' }, size: 10 };
    const fontCell = { bold: false, color: { argb: 'FF111827' }, size: 10 };
    const fontEmpty = { italic: true, color: { argb: 'FF9CA3AF' }, size: 10 };
    const fontDay = { bold: true, color: { argb: 'FF1F2937' }, size: 12 };

    const alignCenter = { horizontal: 'center', vertical: 'middle', wrapText: true };

    // Column A width
    ws.getColumn(1).width = 16;

    // --- Header Row (row 2) ---
    const headerRow = 2;

    // DAY / PERIOD cell
    const dayPeriodCell = ws.getCell(headerRow, 1);
    dayPeriodCell.value = 'DAY / PERIOD';
    dayPeriodCell.font = fontBold;
    dayPeriodCell.alignment = alignCenter;
    dayPeriodCell.fill = fillDay;
    dayPeriodCell.border = borderAll;

    // Period headers
    for (let i = 0; i < periods.length; i++) {
        const p = periods[i];
        const col = i + 2;

        // Find time range from Monday slot
        const slot = slots.find(s => s.period_number === p && s.day_of_week === 'Monday');
        const timeStr = slot ? `${slot.start_time} - ${slot.end_time}` : '';
        const headerText = `Period ${p}\n(${timeStr})`;

        const cell = ws.getCell(headerRow, col);
        cell.value = headerText;
        cell.font = fontHeaderPeriod;
        cell.alignment = alignCenter;
        cell.fill = fillHeader;
        cell.border = borderAll;

        ws.getColumn(col).width = 24;
    }

    // --- Data Rows ---
    let currentRow = headerRow + 1;
    for (const day of activeDays) {
        // Day label cell
        const dayCell = ws.getCell(currentRow, 1);
        dayCell.value = day.substring(0, 3).toUpperCase();
        dayCell.font = fontDay;
        dayCell.alignment = alignCenter;
        dayCell.fill = fillDay;
        dayCell.border = borderAll;

        ws.getRow(currentRow).height = 85;

        // Period cells
        for (let i = 0; i < periods.length; i++) {
            const p = periods[i];
            const col = i + 2;
            const key = `${day}|${p}`;
            const cellEntries = timetableDict[key] || [];

            const cell = ws.getCell(currentRow, col);

            if (cellEntries.length === 0) {
                cell.value = 'Empty Slot';
                cell.fill = fillEmpty;
                cell.font = fontEmpty;
            } else {
                const isLab = cellEntries.some(e => e.session_type && e.session_type.toUpperCase() === 'LAB');
                const lines = cellEntries.map(e => {
                    return `${e.course_code}\n${e.course_name || ''}\n(${e.faculty_name || 'Unassigned'})\n[${e.venue_name || ''}]`;
                });
                cell.value = lines.join('\n---\n');
                cell.fill = isLab ? fillLab : fillClass;
                cell.font = fontCell;
            }

            cell.alignment = alignCenter;
            cell.border = borderAll;
        }

        currentRow++;
    }

    // Generate buffer
    const buffer = await wb.xlsx.writeBuffer();
    return Buffer.from(buffer);
}

module.exports = { generateTimetableExcelBytes };
